bin/recategorize-messages.py

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python
+
+import datetime
+from collections.abc import Iterable
+from typing import Literal
+
+import click
+from sqlalchemy.orm import Query
+
+from inbox.mailsync.backends.imap.common import update_message_metadata
+from inbox.models.account import Account
+from inbox.models.backends.imap import ImapUid
+from inbox.models.folder import Folder
+from inbox.models.message import Message
+from inbox.models.namespace import Namespace
+from inbox.models.session import global_session_scope, session_scope
+
+AccountType = Literal["gmail", "generic", "outlook"]
+ALL_ACCOUNT_TYPES = frozenset({"gmail", "generic", "outlook"})
+
+
+def get_total_namespace_count(
+    *, only_account_id: int | None, only_types: set[AccountType] = ALL_ACCOUNT_TYPES
+) -> int:
+    discriminators = {account_type + "account" for account_type in only_types}
+    namespace_query = (
+        Query([Namespace])
+        .join(Namespace.account)
+        .filter(Account.discriminator.in_(discriminators))
+    )
+    if only_account_id:
+        namespace_query = namespace_query.filter(
+            Namespace.account_id == only_account_id
+        )
+
+    with global_session_scope() as session:
+        return namespace_query.with_session(session).count()
+
+
+def yield_account_id_and_message_ids(
+    *,
+    only_account_id: int | None,
+    date_start: datetime.date | None,
+    date_end: datetime.date | None,
+    only_inbox: bool,
+    only_types: set[AccountType] = ALL_ACCOUNT_TYPES,
+) -> Iterable[int, list[int]]:
+    discriminators = {account_type + "account" for account_type in only_types}
+    namespace_query = (
+        Query([Namespace.account_id, Namespace.id])
+        .join(Namespace.account)
+        .filter(Account.discriminator.in_(discriminators))
+    )
+    if only_account_id:
+        namespace_query = namespace_query.filter(
+            Namespace.account_id == only_account_id
+        )
+
+    with global_session_scope() as session:
+        account_id_to_namespace_id = {
+            account_id: namespace_id
+            for account_id, namespace_id in namespace_query.with_session(session)
+        }
+
+    for account_id, namespace_id in account_id_to_namespace_id.items():
+        query = Query([Message.id]).filter(Message.namespace_id == namespace_id)
+
+        if only_inbox:
+            inbox_folder = ImapUid.folder.has(Folder._canonical_name == "INBOX")
+            query = query.filter(Message.imapuids.any(inbox_folder))
+        if date_start:
+            query = query.filter(Message.created_at >= date_start)
+        if date_end:
+            query = query.filter(Message.created_at < date_end)
```

Comment:
Should we make it inclusive? Given a date range 2024-10-15 - 2024-10-16, I would imagine it's a 2 day interval.

Reply:
Yes, I understand your thinking. I tend to make it exclusive on one side because this way if you run two times [1, 2), [2, 3) you don't reprocess the intersection twice, but in this case we have a single interval.

[`3fc5c0a` (#977)](https://github.com/closeio/sync-engine/pull/977/commits/3fc5c0a0ca653ab5b386d094f9a99e48d5b962d7)